### ISSUE TITLE
Architectural project changes

### ISSUE THREAD

**Issue by SlNPacifist** [ziahamza/webui-aria2#204]
Hi ziahamza!
You recommend asking questions here, so here is my story. I plan to build chrome extension based on webui-aria2 project. It has many advantages compared to running local version of ui or hosting it somewhere.
However, I met several things I'd like to fix:

 * cookies are used as a data storage. This is bad since cookies are sent to server (if ui is hosted somewhere), cookies have limitations, cookies do not work for files and extensions. I want to pick out utils.setCookie and utils.getCookie into a separate module 'storage', which can be replaced when needed.
 * services are prefixed with '$'. This is bad since they could overlap with built-in services. Community name conventions are described [here](https://github.com/johnpapa/angular-styleguide/blob/master/a1/README.md#factory-and-service-names). I want to rename all services into something else.
 * connection settings are stored in a weird way. They are only recorded on successful rpc connect, which prevents saving settings when aria2 server is inaccessible. Moreover, if I have aria2 running on my pc and somewhere else, when remote aria2 server gets down, my settings will be automatically changed to localhost, which is definitely not what I want. I want to change this into more user-friendly way (like automatic server discovery if it is not set directly in settings).
 * Used versions of packages are pretty old. I want to update these packages.

Which of these changes (made as pull requests) you can potentially accept? I plan to make pull request for each change, but not sure if you want to accept those pull requests.

**Comment by ziahamza** [comment]
A Chrome extension would be super helpful! I would personally use it myself!

I completely agree with upgrading old libraries. Moving the storage away from cookies to something like local storage would be cool as well. 

The connection settings are tricky to get right. A check box in connection settings for automatic server discovery would be a nice to have. So by default its checked, but perhaps when you enter connection settings then it unchecked itself. 

Good luck implementing PRs! And feel free to ask questions if anything in the codebase is confusing/not obvious on first sight :)

**Comment by ziahamza** [comment]
So TLDR: Your changes sound very reasonable. I Would be happy to merge them into master

**Comment by SlNPacifist** [comment]
Nice to read that!
TLDR: you can try WebUI packed as an extension [here] (https://github.com/SlNPacifist/aria2-chrome-extension).

WebUI seems to be really easy to implement as an extension. Since I fixed JSONP transport (see pull request hotfix) it connects to my remote server using websocket. UI works pretty well, I found only one glitch for now.
Moving from cookies to localStorage was done without problems. Using chrome.storage API (allows synchronization between instances of chrome), however, requires more work: it is asynchronous. This change allowed connection settings to persist (cookies are not saved for extensions).
Currently working on client-side part of Aria2 RPC. I want to make it a component outside of angular component system so that it can be used for other tasks (like adding new download from context menu).

Closing this issue since I got answers for my questions :)
